[13.0][MIG] Migration of module account_operating_unit

alan196 opened the pull request:
This PR depends on #228

sudhir-serpentcs committed:
[ADD] Added Account Operating Unit module for v9.

JordiBForgeFlow committed:
[FIX] various fixes to ensure that unit tests pass

JordiBForgeFlow committed:
[IMP] changed Eficent copyright

serpentcs-dev1 committed:
[IMP] : Migrated the code to version 9

serpentcs-dev1 committed:
[IMP] Migrated and tested account_ou module to v9.

JordiBForgeFlow committed:
[FIX] travis errors

JordiBForgeFlow committed:
[FIX] remove period_id

JordiBForgeFlow committed:
[FIX] Applied review comments by @elicoidal

JordiBForgeFlow committed:
[FIX] errors in creation of account move

JordiBForgeFlow committed:
[FIX] README.RST style

JordiBForgeFlow committed:
[FIX] travis errors

JordiBForgeFlow committed:
[FIX] travis

JordiBForgeFlow committed:
[IMP] add invoice analysis by operating unit

AaronHForgeFlow committed:
[FIX] Do not assert_balance until all journal items exist. Fix search view in report. Other minor fixes.

AaronHForgeFlow committed:
[FIX] Style fixes suggested by @elicoidal

AaronHForgeFlow committed:
[FIX] travis issues

AaronHForgeFlow committed:
[FIX] travis issues

AaronHForgeFlow committed:
[UPT] By default the OU is not self balanced because the constraint of setting an inter-OU accounts triggers  from other modules when creating companies

AaronHForgeFlow committed:
[UPT] suggested type for inter OU account

JordiBForgeFlow committed:
[FIX] improve README. Inter OU clearing is a balance sheet account

JordiBForgeFlow committed:
[IMP] Add the operating units in the financial reports qweb

AaronHForgeFlow committed:
[FIX] travis issues

JordiBForgeFlow committed:
[IMP] various additions, including:
* Add management of payments using operating units
* Trial Balance by operating units
* Improve tests

JordiBForgeFlow committed:
[IMP] simplify headers

JordiBForgeFlow committed:
[IMP] update readme

JordiBForgeFlow committed:
[FIX] PEP8 issues

JordiBForgeFlow committed:
[IMP] add security for account journals

JordiBForgeFlow committed:
[IMP] add security by OU on the account invoice report

JordiBForgeFlow committed:
[FIX] flake8

pedrobaeza committed:
[MIG] Make modules uninstallable

pedrobaeza committed:
[MIG] Rename manifest files

LoisRForgeFlow committed:
[MIG] account_operating_unit to v10.0

AaronHForgeFlow committed:
[FIX] Company view

AaronHForgeFlow committed:
[FIX] Test remove workflow

JordiBForgeFlow committed:
[10.0][account_operating_unit]Add support for cash basis. Migrate to v10 nomenclature (#55)

* Add support for cash basis. Migrate to v10 nomenclature

* Remove apply_taxes from the create method, as in v10 this parameter is not accepted in the created method

JordiBForgeFlow committed:
Add dependency on analytic_operating_unit. (#62)

The analytic account to be selected in customer
or supplier invoices must have the operating unit
of the invoice as an allowed operating unit.

LoisRForgeFlow committed:
[FIX] fix wrong iteration.

AaronHForgeFlow committed:
[FIX]When paying, the OU of the invoice should be considered only when a single invoice is being paid

serpentcs-dev1 committed:
[FIX] readonly on OU

Commit:
[FIX][account_payment]Fix in the internal transfers amount currency

sbidoul committed:
Allow selecting OU on all account journals

Closes #118

AaronHForgeFlow committed:
[FIX]PYLINT mainly

Commit:
[MIG]account_operating_unit to v11

Commit:
Remove erroneous extra currency code.

To my understanding - the currency lines in liquidity (_get_dst_liquidity_aml_dict_vals) are just wrong and should not be here at all. I understand the need to refactor to get operating unit in, but base Odoo does not have this code about currency.
Now, this fails if the currency and the company currency do not match the journal / target currency.
E.g. A company in Australian Currency transfers 100 Euro from a Euro account to a USD account. When writing to the USD side, self.currency_id will be Euro and self.company_id.currency_id will be AUD, but the transaction will be expecting the USD equivalent.
Base Odoo is currenly handling it all correctly in '_get_shared_move_line_vals', but this code breaks it.

grindtildeath committed:
Changes according to reviews on #114

grindtildeath committed:
Fix default operating_unit_id on account.move.line

bjeficent committed:
[MIG] account_operating_unit: Migration to v12

bjeficent committed:
[FIX] Counterpart's operating unit is wrong and Journals filtered by OUs

jose-pcg committed:
[FIX] account_operating_unit: wrong account.move.line method name calling

oca-travis committed:
[UPD] Update account_operating_unit.pot

OCA-git-bot committed:
[UPD] README.rst

MiquelRForgeFlow committed:
[FIX] Alignments

oca-travis committed:
[UPD] Update account_operating_unit.pot

nikul-serpentcs committed:
[FIX] Fixed unknown Fields

OCA-git-bot committed:
account_operating_unit 12.0.1.0.1

alan196 commented:
I make the migration of this module in this PR.

This module now installs and it seems to work as expected. I only have a question. If someone could help me to answer.

- I see that in the `post` method of `account.move` it has the following code.

```python
    def post(self):
        ml_obj = self.env["account.move.line"]
        for move in self:
            if not move.company_id.ou_is_self_balanced:
                continue

            # If all move lines point to the same operating unit, there's no
            # need to create a balancing move line
            ou_list_ids = [
                line.operating_unit_id and line.operating_unit_id.id
                for line in move.line_ids
                if line.operating_unit_id
            ]
            if len(ou_list_ids) <= 1:
                continue
```

But it also has the following constraint.

```python
    @api.constrains("operating_unit_id", "move_id")
    def _check_move_operating_unit(self):
        for rec in self:
            if (
                rec.move_id
                and rec.move_id.operating_unit_id
                and rec.operating_unit_id
                and rec.move_id.operating_unit_id != rec.operating_unit_id
            ):
                raise UserError(
                    _(
                        "Configuration error. The Operating Unit in"
                        " the Move Line and in the Move must be the"
                        " same."
                    )
                )
```

So there is no way you can get two `account.move.line` with different `operating_unit_id` in the same account move.
The self-balance move line will never be created.

I'm doing something wrong?

AaronHForgeFlow commented:
@alan196 For automatic account moves (almost all) the operating unit will be only in the account move lines.  

The only case where you will find the operating unit  in the account move is when the user creates a manual journal entry. And this is only for the default value when adding lines to the journal entry. So the user puts default  operating unit in the account move and then all the account move lines will be defaulted to that operating unit.

If the manual entry contains account move lines for different operating units then the operating unit in the account move must be blank.

osi-scampbell committed:
[IMP] Ensure Company Set

OCA-git-bot committed:
account_operating_unit 12.0.1.0.2

BT-nstuder commented:
Hi @alan196, is the PR still `WIP`? If yes, are you going to update it?
Regards

AaronHForgeFlow commented:
@BT-nstuder  It seems it is WIP, there are still many references to the deprecated `account.invoice` model.

alan196 commented:
@BT-nstuder  
The migration functionality works (It may need some testing).
The only thing that is pending are the tests, that's the reason you see references of account.invoice

I had a charge of work that doesn't allow me to continue with this migration until next week (February 5 to be more precise).

I will continue with this migration on that date.
If you are interested I can give you access to write in that repo and make updates.

alan196 commented:
@AaronHForgeFlow @BT-nstuder I have adapted the tests and it is ready to be reviewed.

Could you help me to review this PR?

Thanks in advance

AaronHForgeFlow reviewed (changes requested):
I configured the company to have balanced operating units and I created an invoice for Main Operating unit.

Then I paid the invoice using a Bank for a different operating unit. However, the inter-OU accounting entries were not created.

alan196 commented:
> I configured the company to have balanced operating units and I created an invoice for Main Operating unit.
> 
> Then I paid the invoice using a Bank for a different operating unit. However, the inter-OU accounting entries were not created.

@AaronHForgeFlow I make tests to validate this functionality 👍 now that is fixed and with tests.

I added more tests to the payment.

- [x]  Pay two invoices with payment wizard, odoo make a payment per invoice.
- [x] Make a transfer payment between banks in different OU
- [ ] Pay two invoices with payment wizard and group payment.

The latest test is created but the logic to comply with it is still missing, at this moment I'm stuck.
When a payment is done with several invoices of the same OU is not a problem.
But the problems that I define in the test is one invoice from OU1 another from OU2 paid from a bank of OU3.
I imagine two options:
- Split the account move line of the payment for each invoice with their corresponding operating unit and invoice amount. (this must consider multicurrency).
- Raise an error and don't allow the payment of invoices from different OU (I don't think it is a good option).

There is a similar case for invoices from different OU, if you will create a write-off to that payment, what OU needs to be defined in that line???

alan196 commented:
@ougc27

ps-tubtim commented:
@alan196 Please check Travis.

AaronHForgeFlow commented:
Sorry @alan196  did not see your comment above.

> I imagine two options:
Split the account move line of the payment for each invoice with their corresponding operating unit and invoice amount. (this must consider multicurrency).
Raise an error and don't allow the payment of invoices from different OU (I don't think it is a good option).
There is a similar case for invoices from different OU, if you will create a write-off to that payment, what OU needs to be defined in that line???

IMO I think option 1 (split the payment) is the correct one. Nevertheless this options is not implemented in previous version, so in order to move forward, maybe we can include this in the roadmap, and for now, just raise an error and don'nt let create single payments for several different OUs.

Regarding the write-off OU, I think it should go to the OU of the invoice not the payment, but I need to investigate more. 

Thanks for the work!

alan196 commented:
> Sorry @alan196 did not see your comment above.
> 
> > I imagine two options:
> > Split the account move line of the payment for each invoice with their corresponding operating unit and invoice amount. (this must consider multicurrency).
> > Raise an error and don't allow the payment of invoices from different OU (I don't think it is a good option).
> > There is a similar case for invoices from different OU, if you will create a write-off to that payment, what OU needs to be defined in that line???
> 
> IMO I think option 1 (split the payment) is the correct one. Nevertheless this options is not implemented in previous version, so in order to move forward, maybe we can include this in the roadmap, and for now, just raise an error and don'nt let create single payments for several different OUs.
> 
> Regarding the write-off OU, I think it should go to the OU of the invoice not the payment, but I need to investigate more.
> 
> Thanks for the work!

@AaronHForgeFlow 
I agree to continue with the current workflow, but I'm against a raise, what do you think if I add a system parameter, to allow or not the payment with different OU. And document it in Installation instructions.

newtratip commented on account_operating_unit/README.rst:
```diff
@@ -0,0 +1,141 @@
+===============================
```
Could you generate readme to 13.0 ?

newtratip commented on account_operating_unit/models/res_company.py:
```diff
@@ -0,0 +1,36 @@
+    _inherit = "res.company"
+
+    inter_ou_clearing_account_id = fields.Many2one(
+        "account.account",
```
```suggestion
        comodel_name="account.account",
```

newtratip commented on account_operating_unit/models/res_company.py:
```diff
@@ -0,0 +1,36 @@
+    inter_ou_clearing_account_id = fields.Many2one(
+        "account.account",
+        "Inter-operating unit\
+                                                   clearing account",
```
string="Inter-operating unit clearing account"

ps-tubtim replied:
```suggestion
        "Inter-operating unit clearing account",
```

newtratip commented on account_operating_unit/models/res_company.py:
```diff
@@ -0,0 +1,36 @@
+        "Inter-operating unit\
+                                                   clearing account",
+    )
+
```
Remove empty line.

newtratip commented on account_operating_unit/models/res_company.py:
```diff
@@ -0,0 +1,36 @@
+    )
+
+    ou_is_self_balanced = fields.Boolean(
+        "Operating Units are self-balanced",
```
```suggestion
        string="Operating Units are self-balanced",
```

ps-tubtim commented on account_operating_unit/models/account_move.py:
```diff
@@ -0,0 +1,231 @@
+                move = self.env["account.move"].browse(vals["move_id"])
+                if move.operating_unit_id:
+                    vals["operating_unit_id"] = move.operating_unit_id.id
+        return super(AccountMoveLine, self).create(vals_list)
```
```suggestion
        return super().create(vals_list)
```

ps-tubtim commented on account_operating_unit/models/account_move.py:
```diff
@@ -0,0 +1,231 @@
+            domain.append(
+                ("operating_unit_id", "in", self._context.get("operating_unit_ids"))
+            )
+        return super(AccountMoveLine, self)._query_get(domain)
```
```suggestion
        return super()._query_get(domain)
```

ps-tubtim commented on account_operating_unit/models/account_move.py:
```diff
@@ -0,0 +1,231 @@
+                    )
+                )
+
+    @api.constrains("operating_unit_id", "move_id")
+    def _check_move_operating_unit(self):
+        for rec in self:
+            if (
+                rec.move_id
+                and rec.move_id.type == "entry"
+                and rec.move_id.operating_unit_id
+                and rec.operating_unit_id
+                and rec.move_id.operating_unit_id != rec.operating_unit_id
+            ):
+                raise UserError(
+                    _(
+                        "Configuration error. The Operating Unit in"
+                        " the Move Line and in the Move must be the"
+                        " same."
+                    )
+                )
```
IMO, I think you should check all move types. Why you check `rec.move_id.type == "entry"`?

alan196 replied:
This test is only valid when you add the account move manually, as commented by @AaronHForgeFlow in this comment https://github.com/OCA/operating-unit/pull/229#issuecomment-570145352 that's the reason this constraint only validate moves when are of type `entry`

ps-tubtim commented on account_operating_unit/models/account_move.py:
```diff
@@ -0,0 +1,231 @@
+class AccountMove(models.Model):
+    _inherit = "account.move"
+
+    @api.model
+    def _default_operating_unit_id(self):
+        if (
+            self._context.get("default_type", False)
+            and self._context.get("default_type") != "entry"
+        ):
+            return self.env["res.users"].operating_unit_default_get()
+        return False
+
+    operating_unit_id = fields.Many2one(
+        comodel_name="operating.unit",
+        default=_default_operating_unit_id,
```
You can use `default=lambda self: self.env["res.users"].operating_unit_default_get(),` instead this code.

alan196 replied:
This is related with the comment https://github.com/OCA/operating-unit/pull/229/files#r408973265

You don't need a default operating unit to be defined when the type is entry

ps-tubtim commented on account_operating_unit/models/account_move.py:
```diff
@@ -0,0 +1,231 @@
+        for line in self.line_ids:
+            if self.operating_unit_id:
+                line.operating_unit_id = self.operating_unit_id
+        return res
```
```suggestion
    @api.onchange("invoice_line_ids")
    def _onchange_invoice_line_ids(self):
        if self.operating_unit_id:
            for line in self.line_ids:            
                line.operating_unit_id = self.operating_unit_id
        return super()._onchange_invoice_line_ids()
```

ps-tubtim commented on account_operating_unit/models/account_move.py:
```diff
@@ -0,0 +1,231 @@
+                    {"line_ids": [(4, aml.id) for aml in amls]}
+                )
+
+        return super(AccountMove, self).post()
```
```suggestion
        return super().post()
```

ps-tubtim commented on account_operating_unit/models/account_move.py:
```diff
@@ -0,0 +1,231 @@
+    def _check_balanced(self):
+        if self.env.context.get("wip"):
+            return True
+        return super(AccountMove, self)._check_balanced()
```
```suggestion
        return super()._check_balanced()
```

ps-tubtim commented on account_operating_unit/models/account_payment.py:
```diff
@@ -0,0 +1,42 @@
+    )
+
+    def _prepare_payment_moves(self):
+        res = super(AccountPayment, self)._prepare_payment_moves()
```
```suggestion
        res = super()._prepare_payment_moves()
```

ps-tubtim commented on account_operating_unit/report/account_invoice_report.py:
```diff
@@ -0,0 +1,26 @@
+    operating_unit_id = fields.Many2one("operating.unit", "Operating Unit")
+
+    def _select(self):
+        select_str = super(AccountInvoiceReport, self)._select()
```
```suggestion
        select_str = super()._select()
```

ps-tubtim commented on account_operating_unit/report/account_invoice_report.py:
```diff
@@ -0,0 +1,26 @@
+        return select_str
+
+    def _group_by(self):
+        group_by_str = super(AccountInvoiceReport, self)._group_by()
```
```suggestion
        group_by_str = super()._group_by()
```

newtratip commented on account_operating_unit/views/account_payment_view.xml:
```diff
@@ -0,0 +1,69 @@
+                    groups="operating_unit.group_multi_operating_unit"
+                />
+            </field>
+            <field name="journal_id" position="after">
```
```suggestion
            
```

newtratip commented on account_operating_unit/views/account_payment_view.xml:
```diff
@@ -0,0 +1,69 @@
+                    name="operating_unit_id"
+                    groups="operating_unit.group_multi_operating_unit"
+                />
+            </field>
```
```suggestion
            
```

newtratip commented on account_operating_unit/views/account_payment_view.xml:
```diff
@@ -0,0 +1,69 @@
+            </field>
+        </field>
+    </record>
+    <record id="view_account_supplier_payment_tree" model="ir.ui.view">
```
Supplier Payment Tree is show 2 columns for Operating Unit. You don't need to add this record because view_account_supplier_payment_tree inherited view_account_payment_tree which have column Operating Unit already in record above.
![Selection_118](https://user-images.githubusercontent.com/24691983/77915916-7af01e00-72c2-11ea-814f-22c1ac9067e4.png)

newtratip reviewed (commented):
Some Review.

newtratip commented on account_operating_unit/views/account_move_view.xml:
```diff
@@ -0,0 +1,85 @@
+        <field name="model">account.move.line</field>
+        <field name="inherit_id" ref="account.view_account_move_line_filter" />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='account_id']" position="after">
```
```suggestion
            
```

newtratip commented on account_operating_unit/views/account_move_view.xml:
```diff
@@ -0,0 +1,85 @@
+                        groups="operating_unit.group_multi_operating_unit"
+                    />
+                </filter>
+            </xpath>
```
```suggestion
            
```

newtratip commented on account_operating_unit/views/account_move_view.xml:
```diff
@@ -0,0 +1,85 @@
+                        groups="operating_unit.group_multi_operating_unit"
+                    />
+                </field>
+                <filter name="account" position="after">
```
```suggestion
                <filter name="group_by_account" position="after">
```

newtratip commented on account_operating_unit/views/account_move_view.xml:
```diff
@@ -0,0 +1,85 @@
+                    'line_ids': line_ids,
+                    'journal_id': journal_id,
+                    'default_operating_unit_id': operating_unit_id}
+                </attribute>
```
I think context should copy from original view and add `'default_operating_unit_id': operating_unit_id` at last this.
![Selection_119](https://user-images.githubusercontent.com/24691983/77944144-9ff78780-72e8-11ea-9cce-7ef241726535.png)

newtratip commented:
@alan196 Could you add Operating Unit in tree view and search view of invoice ?

newtratip commented on account_operating_unit/views/account_move_view.xml:
```diff
@@ -0,0 +1,85 @@
+        <field name="inherit_id" ref="account.view_move_form" />
+        <field name="arch" type="xml">
+            <field name="journal_id" position="after">
+                <field name="operating_unit_id" />
```
```suggestion
                <field name="operating_unit_id" options="{'no_create': True}" groups="operating_unit.group_multi_operating_unit" />
```

newtratip commented on account_operating_unit/security/account_security.xml:
```diff
@@ -0,0 +1,58 @@
+<?xml version="1.0" encoding="utf-8" ?>
```
Add rule for account move too.

newtratip commented on account_operating_unit/report/account_invoice_report.py:
```diff
@@ -0,0 +1,26 @@
+
+    _inherit = "account.invoice.report"
+
+    operating_unit_id = fields.Many2one("operating.unit", "Operating Unit")
```
```suggestion
    operating_unit_id = fields.Many2one(comodel_name="operating.unit", string="Operating Unit")
```

newtratip commented on account_operating_unit/models/account_payment.py:
```diff
@@ -0,0 +1,42 @@
+                payment.operating_unit_id = payment.journal_id.operating_unit_id
+
+    operating_unit_id = fields.Many2one(
+        "operating.unit",
```
```suggestion
        comodel_name="operating.unit",
```

newtratip commented on account_operating_unit/models/res_company.py:
```diff
@@ -0,0 +1,36 @@
+        "operating unit.",
+    )
+
+    @api.constrains("ou_is_self_balanced")
```
```suggestion
    @api.constrains("ou_is_self_balanced", "inter_ou_clearing_account_id")
```
IMO, If you change inter_ou_clearing_account_id to empty and ou_is_self_balanced still check. The system should show raise error too.

AaronHForgeFlow reviewed (commented):
The inter operating unit account entries are created correctly when paying invoices from different OU. For me it is ok to move forward like it is now. Just pending the CI checks to proceed to merge.

AaronHForgeFlow reviewed (approved):
:+1:  Functional tested. Thank you!

lmgr0312 reviewed (approved):
LGTM!

AaronHForgeFlow commented:
Will merge when comments are attended.

alan196 committed:
[IMP] account_operating_unit: black, isort, prettier

alan196 committed:
[MIG] account_operating_unit: Migration to V13

AaronHForgeFlow commented:
All comments attended. Thank you @alan196 !

AaronHForgeFlow commented:
/ocabot merge
